rfcs/20230801-max-partition-policy/README.md

```diff
@@ -0,0 +1,148 @@
+# Introducing Max Partition Policy to oneDNN Graph API
+
+## Motivation
+
+oneDNN Graph Compiler is currently a experimental backend for oneDNN Graph API.
+It provides
+optimized implementations for complex computational graphs including multi-head
+attention (MHA), multi-layer perceptron (MLP), and convolution residual blocks
+over typical data types for both inference and training.
+
+While the predefined patterns meet some user requirements, there is a growing
+interest in customized ops and patterns, which the current fixed patterns
+cannot fully accommodate. For instance, users may want to build RMSNorm with
+small ops and expect the library to provide optimizations for it.
+This is a frequent request for newly introduced activation or normalization.
+
+To address this need and provide users with more flexible and efficient options,
+oneDNN Graph Compiler is seeking to incorporate a max partition policy into oneDNN
+Graph API, enabling direct programming and unlocking further performance optimizations.
+
+## Current Status
+
+In oneDNN Graph API programming model, a computation graph is passed to library
+and then optimized partitions are returned by the library by
+`graph.get_partitions()` API.
+
+```c
+dnnl_status_t dnnl_graph_graph_filter(
+        dnnl_graph_graph_t graph, dnnl_graph_partition_policy_t policy);
+```
+
+```cpp
+class graph {
+    ...
+    std::vector<partition> get_partitions(
+            partition::policy policy = partition::policy::fusion);
+    ...
+};
+```
+
+A partition policy is used in the API to control granularities
+of optimizations to be applied when partitioning.
+Currently, oneDNN Graph API supports two partition policies:
+
+```c
+typedef enum {
+    dnnl_graph_partition_policy_fusion = 1,
+    dnnl_graph_partition_policy_debug = 2,
+} dnnl_graph_partition_policy_t;
+```
+
+```cpp
+class partition {
+    ...
+    enum class policy {
+        fusion = dnnl_graph_partition_policy_fusion,
+        debug = dnnl_graph_partition_policy_debug,
+    };
+    ...
+};
+```
+
+- Fusion policy (default partition policy): fusion policy returns partitions
+  with [pre-defined fusion patterns](https://oneapi-src.github.io/oneDNN/dev_guide_graph_fusion_patterns.html),
+  e.g. Convolution followed by a chain of post-ops, MHA, MLP, etc.
+- Debug policy: debug policy returns partitions with single operation in each partition.
+  This policy doesn't apply any fusions, which is useful when
+  user wants to debug or compare performance with fusion policy.
+
+## Proposal
+
+As mentioned in previous sessions, fusion policy relies on predefined fusion
+patterns to do partitioning, which may miss some optimization opportunities
+for customized ops and patterns that have not been predefined.
+To address this, max policy is proposed.
+
+The expected behaviors when using max policy:
+
+- Max policy relies on predefined ops list to do partitioning. Connected ops
```

Comment:
Choice between `max` and `fusion` policy falls on the shoulders of the framework user. How this is expected to be exposed in Pytorch and Tensorflow?

How users are supposed to decide whether use fusion or max? The explanation here talks about implementation details mainly, but what are the consequences visible to framework user?

Reply:
My understanding is,

`Max` policy is not expected to be exposed by framework, instead, it's used to support framework developer to implement their custom fusion kernels.

`Fusion` policy is the recommended way to use onednn graph automatically, as we will do graph partition according to pre-defined fusion pattern and deliver good performance. While `max` policy is to support direct programming with onednn graph API, and framework developers need to program the graph they want to fuse and pass it to onednn graph with `max` policy, just like how they do C program and compile it with GCC. 

`max` policy can be used in below scenarios:

1. The default `fusion` policy doesn't meet framework's requirement and developer wants to change the fusion behavior, for example, for a 3-layer MLP, `fusion` policy will take them all, but if the developer only wants oneDNN graph to take the first two layers, and leave the last layer to be un-partitioned, then they can program with `max` policy and define a graph only contains first two layers.
2. Define new fusion patterns. For example, if framework developer want to implement RMSNorm op, they can build a graph with small ops and pass it to oneDNN graph with `max` policy, then a fused kernel will be returned.

In general, `max` policy is not designed for framework users, but for framework developers.

Reply:
@Jianhui-Li, adding to your point (custom ops vs. patterns created by max partitioning), I feel like there's another aspect as well - a pattern added manually by backend teams may sometimes perform better than the pattern automatically chosen by the max partitioning approach, and sometimes, the opposite may be true.

 So at runtime, should a framework (transparently to framework users) benchmark both approaches, and eventually use the approach that delivers better performance? In case a framework can't do this sort of benchmarking, then the framework may have to expose an API to the users to manually check if max partitioning approach (if it's non-default) may deliver better performance.